src/Microsoft.ApiDesignGuidelines.Analyzers/Core/ImplementIDisposableCorrectly.cs

```diff
@@ -106,13 +112,494 @@ public abstract class ImplementIDisposableCorrectlyAnalyzer : DiagnosticAnalyzer
+                    if (suppressFinalizeMethod == null)
+                    {
+                        return;
+                    }
+
+                    var analyzer = new Analyzer(context.Compilation, disposableType, disposeInterfaceMethod, suppressFinalizeMethod);
+                    analyzer.Initialize(context);
+                });
+        }
+
+        private static bool IsDisposeBoolMethod(IMethodSymbol method)
+        {
+            if (method.Name == DisposeMethodName && method.MethodKind == MethodKind.Ordinary &&
+                method.ReturnsVoid && method.Parameters.Length == 1)
+            {
+                var parameter = method.Parameters[0];
+                if (parameter.Type != null && parameter.Type.SpecialType == SpecialType.System_Boolean && parameter.RefKind == RefKind.None)
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Analyzes single instance of compilation.
+        /// </summary>
+        private class Analyzer
+        {
+            private readonly Compilation compilation;
+            private readonly INamedTypeSymbol disposableType;
+            private readonly IMethodSymbol disposeInterfaceMethod;
+            private readonly IMethodSymbol suppressFinalizeMethod;
+
+            public Analyzer(Compilation compilation, INamedTypeSymbol disposableType, IMethodSymbol disposeInterfaceMethod, IMethodSymbol suppressFinalizeMethod)
+            {
+                this.compilation = compilation;
+                this.disposableType = disposableType;
+                this.disposeInterfaceMethod = disposeInterfaceMethod;
+                this.suppressFinalizeMethod = suppressFinalizeMethod;
+            }
+
+            public void Initialize(CompilationStartAnalysisContext context)
+            {
+                context.RegisterSymbolAction(AnalyzeNamedTypeSymbol, SymbolKind.NamedType);
+                context.RegisterOperationBlockAction(AnalyzeOperationBlock);
+            }
+
+            private void AnalyzeNamedTypeSymbol(SymbolAnalysisContext context)
+            {
+                var type = context.Symbol as INamedTypeSymbol;
+                if (type != null && type.IsType && type.TypeKind == TypeKind.Class)
+                {
+                    var implementsDisposableInBaseType = ImplementsDisposableInBaseType(type);
+
+                    if (ImplementsDisposableDirectly(type))
+                    {
+                        var disposeMethod = FindDisposeMethod(type);
+                        if (disposeMethod != null)
+                        {
+                            // This is difference from FxCop implementation
+                            // IDisposable Reimplementation Rule is violated only if type re-implements Dispose method, not just interface
```

Comment:
Yes, the deviation addresses false-positive from FxCop. For example FxCop reports warning on class C in following code:
```
    public interface ITest : IDisposable
    {
        void Test();
    }

    public class B : IDisposable
    {
        public void Dispose()
        {
            Dispose(true);
            GC.SuppressFinalize(this);
        }

        protected virtual void Dispose(bool disposing)
        {
        }

        ~B()
        {
            Dispose(false);
        }
    }

    public class C : B, ITest
    {
        public void Test()
        {
        }
    }
```

However, it is obvious that C does not re-implement IDisposable and there is no way to remove IDisposable from type C.